asynq/tools.py

```diff
@@ -305,3 +307,39 @@ def resume(self):
 
     def pause(self):
         self.total_time += utime() - self._last_start_time
+
+
+class AsyncEventHook(EventHook):
+    """EventHook that supports async handlers.
+
+    When the event triggers, all the handlers will be invoked asynchronously using async_call.
+
+    Since it uses async_call, it supports normal handlers by default
+    (they don't need to be @async() decorated).
+
+    """
+
+    @async()
+    def trigger(self, *args):
+        yield [async_call.async(handler, *args) for handler in self]
+
+    @async()
+    def safe_trigger(self, *args):
+        wrapped_handlers = [self._create_safe_wrapper(handler) for handler in self]
+        results = yield [wrapped_handler.async(*args) for wrapped_handler in wrapped_handlers]
+        results = filter(None, results)
+        if len(results) > 0:
+            reraise(results[0])
+
+    @staticmethod
+    def _create_safe_wrapper(handler):
+        @async()
+        def wrapped(*args):
+            error = None
+            try:
+                yield async_call.async(handler, *args)
+            except BaseException as e:
```

Comment:
I'm uncomfortable with catching BaseException here because we may end up eating a KeyboardInterrupt if the first handler throws a normal exception, the second throws KeyboardInterrupt, and we reraise the first one. qcore has the same issue though, so let's discuss it there first and then revisit this.